### DISCUSSION TITLE
How does versioning works under the hood?

### DISCUSSION THREAD

**Discussion by HeldvonKosmos** [git-lfs/git-lfs#5910] [General]
I understood, that it is working with pointers, but will a whole copy of the file made, when it changes?
Isn't that very inefficient in terms of spacecomplexity.
A good approach is like the Backup-Tool "Bup"
It uses the deltaalgorithm of rsync.
And then the different versions can only contain the specific changes that are made.

**Comment by chrisd8088** [comment]
Hey, thanks for the interesting question!

I've tried to reply in some detail below, but in brief, this isn't something we are likely to pursue in this project.

However, if someone wants to create a PR which includes a detailed implementation plan with appropriate API extensions in a new document in our `/docs/proposals` [directory](https://github.com/git-lfs/git-lfs/tree/main/docs/proposals), we would of course be pleased to take a look at it!

#### Git Data Model

Git's basic [data model](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects) consists of Git objects, which may be commits, trees, or blobs, with each commit pointing to a tree, trees containing blobs and other trees, and blobs containing file data.  Commits also point backwards to prior parent commits.  One consequence of this model is that blobs are complete files, and each commit represents a complete snapshot of the repository.

Git's "porcelain" commands, like `git log -p`, do a good job of presenting Git's history as a series of changes ("diffs" or "deltas"), but that's kind of an illusion; really, each commit is a complete representation of all the files in the repository, and Git has to compute the diffs it shows to a user.  (There are alternative source code management tools like Pijul which [store](https://pijul.org/manual/why_pijul.html#pijul-for-gitmercurialsvn-users) their history as a series of patches, not snapshots.) 

#### Git Attributes

The basic principles of Git's data storage model are reflected in the way Git handles [attributes](https://git-scm.com/book/ms/v2/Customizing-Git-Git-Attributes) and in particular in the [design](https://git-scm.com/docs/gitattributes#_filter) of Git's `filter` attribute mechanism, which is a key point of contact between Git and Git LFS.  When you run `git lfs track "*.jpg"`, you create a `.gitattributes` file with a line like `*.jpg filter=lfs diff=lfs merge=lfs -text`.  The `filter=lfs` attribute is crucial, as it tells Git to invoke the Git LFS client whenever a new `*.jpg` file is checked in or a Git blob representing a `*.jpg` file is checked out into a working tree.

When checking in a file, for instance, the file's contents are sent to a `clean` filter command, which in Git LFS's case is the `git-lfs-clean(1)` command (or more likely, with modern Git versions, the `git-lfs-filter-process(1)` command, which handles multiple files, but the concepts are the same).  The exchange here is effectively as if Git says, "Hey, here's a file, please transform it as you like and then give me the transformed contents to store in my repository".  Conversely, when checking out a file, Git invokes a `smudge` filter command like the `git-lfs-smudge(1)` command, and effectively says, "Hey, here's what I have in my repository for this file; please transform it as you like and then give me the transformed contents to write into the working tree".  In the words of the Git [manual](https://git-scm.com/docs/gitattributes#_filter):

>  Upon checkout, when the `smudge` command is specified, the command is fed the blob object from its standard input, and its standard output is used to update the worktree file.

So the filter commands receive no information about the history of a file or blob, merely the file or blob's contents.  Because Git LFS is designed around this mechanism, its basic model is to either store or retrieve an entire file, and to save or read a pointer to that file as a Git blob, without any knowledge of the history of that file.  When checking in a file, Git LFS calculates the SHA-256 hash of the content of the file (as it receives from Git via the `clean` filter protocol), stores the file, and puts the hash into the pointer it returns as the contents of a blob for Git to store.  When checking out a blob (through the `smudge` filter protocol), if the blob's contents are recognized as a Git LFS pointer, Git LFS retrieves the file which corresponds to the SHA-256 hash in the pointer and passes the file's data on to Git, and otherwise just passes the blob's contents straight through to Git; Git then writes whatever it receives into a file in the working tree.

#### Git LFS Batch API

To store and retrieve the actual file data, Git LFS communicates with a server using the Git LFS Batch API [specification](https://github.com/git-lfs/git-lfs/blob/3990c7ab4fa6c99540af3f73f33da47a72df08af/docs/api/batch.md).  The exchanges here are as if the Git LFS client says, "Hey, please store this file using this SHA-256 hash value as the file's ID", and "Hey, please give me the contents of the file whose ID is this SHA-256 hash value".  Again, there's no point in this protocol where any history of the file is exchanged, and that follows from the way Git LFS is integrated with Git using the `filter` attribute system, and that in turn follows from the way Git models commits and blobs.

#### Git Internal Delta Compression

Now to be fair to Git, it _does_ perform a lot of delta analysis, and indeed uses those deltas to efficiently compact its internal data files!  There's a good [five](https://github.blog/open-source/git/gits-database-internals-i-packed-object-store/)-[part](https://github.blog/open-source/git/gits-database-internals-ii-commit-history-queries/) [series](https://github.blog/open-source/git/gits-database-internals-iii-file-history-queries/) of [blog](https://github.blog/open-source/git/gits-database-internals-iv-distributed-synchronization/) [posts](https://github.blog/open-source/git/gits-database-internals-v-scalability/) which cover how this all works in detail, but one paragraph from the first [article](https://github.blog/open-source/git/gits-database-internals-i-packed-object-store/#diffable-object-content) is especially relevant here:

> Git was built to store source code, which consists of plain-text files that are used as input to a compiler or interpreter to create applications.  Git was also built to store many versions of this source code as it is changed by humans.  This provides additional context about the kind of data typically stored in Git: diffable files with significant portions in common.  If you’ve ever wondered why you shouldn’t store large binary files in Git repositories, this is the reason.

The kinds of advanced optimizations Git can use to efficiently store plain-text files by comparing them and finding deltas start to break down when it is asked to store binary files whose contents are opaque.  Indeed, that's a key reason why major Git hosting services may prevent users from storing Git blobs over a certain size, and encourage them to use tools like Git LFS for this kind of content instead.  

#### Git LFS History

As I mentioned recently in another discussion, the Git LFS project grew out of several concurrent efforts by developers at various Git hosting services who were all faced with these kinds of concerns.  There's an archived [blog post](https://web.archive.org/web/20240302022611/https://blog.developer.atlassian.com/contributing-to-git-lfs/) which talks about this history, and this [video](https://www.youtube.com/watch?v=p3Pse1UkEhI) covers some of the same background as well.

#### Potential Git LFS Optimizations

Now just as Git's abstract data model doesn't track differences between files, but its internal storage optimizations look for those, there's nothing preventing a Git hosting service which offers Git LFS storage from trying to find deltas between Git LFS objects and store those instead of storing complete objects.  Such a service implementation could reference the Git history and look for Git LFS pointer files, and then check if they were new revisions of other Git LFS pointers, and if so compare them and store just the differences it found, along with a chain of references from the older revisions to the newer ones.

I suspect no Git LFS hosting services have attempted to implement such a scheme, however, for a few reasons.  First, it would introduce a lot of complexity and fragility with limited benefits, compared to just storing whole Git LFS objects.  Second, the cost of mass data storage was low even ten years ago, when Git LFS was first developed, and has only fallen further since.  (When I was young, storage was quite expensive and so we used to pay attention to every kilobyte of disk usage, but the historical [trend](https://ourworldindata.org/grapher/historical-cost-of-computer-memory-and-storage) in storage costs makes that kind of optimization much less valuable these days except under specific circumstances.)

Where there could be a real benefit to Git LFS hosting services and users, though, would be if there were ways to reduce the bandwidth consumed by the transfer of Git LFS objects.  That would of course be very much akin to the purpose of rsync's delta algorithm you mentioned, to speed up file transfers by limiting the amount of data which needs to be sent and received.  For users of commercial Git hosting services who have large numbers of Git LFS objects in their repositories, repeatedly fetching those in bulk can drive up the amount they are charged for Git LFS usage much more quickly than just adding more objects into storage, because bandwidth costs are generally significantly higher than storage costs.

Making that possible, though, would require a more advanced version of the Git LFS Batch API, where the Git LFS client and server could negotiate as to what data the client already has and what it needs.  In theory this might work a lot like Git's v2 [transfer protocol](https://git-scm.com/docs/protocol-v2#_fetch) for [fetches](https://git-scm.com/book/en/v2/Git-Internals-Transfer-Protocols#_downloading_data), where lists of `have` and `want` entries are exchanged.  The Git LFS server could say, "Hey, you asked for object A, do you have any of A′, A″, or A‴?", and the client could reply, "Yes, I have A″ and A‴", and then the server could compute just the difference between A and A″ and send that back instead of the entire contents of A.

#### Compressed File Formats

Again, though, to my knowledge no one has proposed such an extension to the Git LFS APIs.  This is likely in part due to the complexity it would add, but I suspect the main reason is that so many binary files of the kinds people store with Git LFS are compressed, and that makes it much harder for a naive algorithm to find significant commonalities between files.  The rsync utility encounters this problem too, as the Wikipedia entry [states](https://en.wikipedia.org/wiki/Rsync#Determining_which_parts_of_a_file_have_changed):

> If the sender's and recipient's versions of the file have many sections in common, the utility needs to transfer relatively little data to synchronize the files.  If typical [data compression](https://en.wikipedia.org/wiki/Data_compression) algorithms are used, files that are similar when uncompressed may be very different when compressed, and thus the entire file will need to be transferred. Some compression programs, such as [gzip](https://en.wikipedia.org/wiki/Gzip), provide a special "rsyncable" mode which allows these files to be efficiently rsynced, by ensuring that local changes in the uncompressed file yield only local changes in the compressed file.

Many of the most common types of binary file types are compressed using a variety of techniques (e.g., JPEGs, MPEGs, PDFs, Microsoft Office formats, etc.)  Sometimes the entire file is compressed, and sometimes only internal portions, but either way, a generic tool like Git LFS or rsync is going to be at a major disadvantage compared to tools dedicated to a specific format in trying to examine their contents and look for common sections between multiple files.  Adding format-specific logic to Git LFS would be a huge undertaking, and would greatly increase the chance of bugs and incompatibilities introducing problems.  No one would want to use a utility like Git LFS if it caused data corruption because it didn't fully understand the details of some specific file format.

By contrast, for applications and services which have intimate knowledge of the data they are transmitting, considerable improvements over raw data transfer are possible.  Just as one example, there are a number of [blog](https://netflixtechblog.com/dynamic-optimizer-a-perceptual-video-encoding-optimization-framework-e19f1e3a277f) [posts](https://netflixtechblog.com/optimized-shot-based-encodes-for-4k-now-streaming-47b516b10bbb) from Netflix which describe how they can analyze video streams on a frame-by-frame and shot-by-shot basis and very efficiently encode the differences to maximize the level of detail they retain while minimizing the size of the data which has to be streamed to clients.

#### Closing Thoughts

So that's my long answer to your very interesting question!  I think this topic touches on a lot of key aspects of both Git and Git LFS's design principles and history.

To reiterate what I said before, if anyone comes up with an novel proposal to try to improve Git LFS storage or bandwidth use and puts it into a PR, we would be pleased to discuss it there!
